src/Microsoft.CodeQuality.Analyzers.Exp/Core/Reliability/DisposeObjectsBeforeLosingScope.cs

```diff
@@ -0,0 +1,141 @@
+                compilationContext.RegisterOperationBlockStartAction(operationBlockStartContext =>
+                {
+                    bool hasDisposableCreation = false;
+                    operationBlockStartContext.RegisterOperationAction(operationContext =>
+                    {
+                        if (!hasDisposableCreation &&
+                            operationContext.Operation.Type.IsDisposable(iDisposable))
+                        {
+                            hasDisposableCreation = true;
+                        }
+                    },
+                    OperationKind.ObjectCreation,
+                    OperationKind.TypeParameterObjectCreation,
+                    OperationKind.DynamicObjectCreation,
+                    OperationKind.Invocation);
+
+                    operationBlockStartContext.RegisterOperationBlockEndAction(operationBlockEndContext =>
+                    {
+                        if (!hasDisposableCreation ||
+                            !(operationBlockEndContext.OwningSymbol is IMethodSymbol containingMethod))
+                        {
+                            return;
+                        }
+
+                        foreach (var operationRoot in operationBlockEndContext.OperationBlocks)
+                        {
+                            IBlockOperation topmostBlock = operationRoot.GetTopmostParentBlock();
+                            if (topmostBlock != null)
+                            {
+                                var cfg = ControlFlowGraph.Create(topmostBlock);
+                                var nullAnalysisResult = NullAnalysis.GetOrComputeResult(cfg, containingMethod.ContainingType);
+                                var pointsToAnalysisResult = PointsToAnalysis.GetOrComputeResult(cfg, containingMethod.ContainingType, nullAnalysisResult);
+                                var disposeAnalysisResult = DisposeAnalysis.GetOrComputeResult(cfg, iDisposable, iCollection,
+                                    genericICollection, disposeOwnershipTransferLikelyTypes, containingMethod.ContainingType, pointsToAnalysisResult, nullAnalysisResult);
+                                ImmutableDictionary<AbstractLocation, DisposeAbstractValue> disposeDataAtExit = disposeAnalysisResult[cfg.Exit].InputData;
+                                foreach (var kvp in disposeDataAtExit)
+                                {
+                                    AbstractLocation location = kvp.Key;
+                                    DisposeAbstractValue disposeValue = kvp.Value;
+                                    if (disposeValue.Kind == DisposeAbstractValueKind.NotDisposed ||
+                                        ((disposeValue.Kind == DisposeAbstractValueKind.Disposed ||
+                                          disposeValue.Kind == DisposeAbstractValueKind.MaybeDisposed) &&
```

Comment:
If you add a `if (disposedValue.Kind == DisposeAbstractValueKind.NotDisposable) continue;` at the beginning, then you don't have to have this part of the conditional at all, which will make it more readable to me.